### PR TITLE
ref(metrics_extraction): Change assert_on_demand_model and typing changes

### PR DESCRIPTION
**Pull request by armenzg**
This is preparatory work for another PR and makes the typing check tighter.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -644,8 +644,10 @@ disable_error_code = [
 # beginning: stronger typing
 [[tool.mypy.overrides]]
 module = [
+    "src.sentry.tasks.on_demand_metrics",
```
Comment (armenzg) [inline pyproject.toml:647]: Adding these two modules in this block prohibits them from having function signatures without fully typing them.

File: tests/sentry/tasks/test_on_demand_metrics.py
Comment (armenzg) [inline tests/sentry/tasks/test_on_demand_metrics.py:1]: The first half of the changes are just typing changes.

File: tests/sentry/tasks/test_on_demand_metrics.py [hunk 2]
```diff
@@ -362,14 +365,23 @@ def test_process_widget_specs(
             for mock_call in _set_cardinality_cache.mock_calls
         )
 
+    expected_state = ""
```
Comment (armenzg) [inline tests/sentry/tasks/test_on_demand_metrics.py:368]: I'm dropping the `expected_applicable` parameter from `assert_on_demand_model` and adding `expected_state`, thus, I move some internal logic to this place in this test.

The test I will be adding in my next PR uses `OnDemandExtractionState.DISABLED_NOT_APPLICABLE` which `assert_on_demand_model` does not support checking for before my changes in here.

File: tests/sentry/tasks/test_on_demand_metrics.py [hunk 3]
```diff
@@ -362,14 +365,23 @@ def test_process_widget_specs(
...
+    if not feature_flags:
+        expected_state = OnDemandExtractionState.DISABLED_PREROLLOUT
+    else:
+        expected_state = (
+            OnDemandExtractionState.ENABLED_ENROLLED
+            if expected_low_cardinality
+            else OnDemandExtractionState.DISABLED_HIGH_CARDINALITY
+        )
```
Comment (armenzg) [inline tests/sentry/tasks/test_on_demand_metrics.py:376]: This block comes from this one:
image

**Review by asottile-sentry — APPROVED** [review]

**Pull request merged**